Refactor -o & --coverage implementation | Fixes #6859

MarcoScabbiolo opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

Added `DependencyResolver#resolveInverseModuleMap` that exposes all the module map (the files and their dependencies). `resolveInverse` was kept in its original form for compatibility.

The changed files are no longer attached to `collectCoverageFrom`. Now they're passed through runners, schedulers and workers to `Runtime.shouldInstrument`, the proper method to exclude files from coverage.

Includes e2e regression tests.

Fixes #6859, fixes #7101, closes #7153

SimenB commented:
Thank you so much! A failing test is the best sort of bug report ❤️

MarcoScabbiolo commented:
Looks like the Windows job had a blue screen :(

SimenB commented:
That's awesome @MarcoScabbiolo, thanks! I added a changelog entry to this.

However, I'm not entirely sure this is the correct fix. For instance, if I add this diff the the integration test, it fails again:

```diff
diff --git i/e2e/__tests__/onlyChanged.test.js w/e2e/__tests__/onlyChanged.test.js
index 8f85d0dd9..9aa64f785 100644
--- i/e2e/__tests__/onlyChanged.test.js
+++ w/e2e/__tests__/onlyChanged.test.js
@@ -151,7 +151,7 @@ test('do not pickup non-tested files when reporting coverage on only changed fil
   run(`${GIT} commit --no-gpg-sign -m "first"`, DIR);
 
   writeFiles(DIR, {
-    'b.test.js': 'it("passes", () => {})',
+    'b.test.js': 'require("./package.json");\nit("passes", () => {})',
     'package.json': JSON.stringify({name: 'new name'}),
   });
 
```

That will _not_ fail if we remove `-o`. We're missing some filter that would remove files we never collect coverage from

---

>Looks like the Windows job had a blue screen :(

Ignore it, see #7760

MarcoScabbiolo commented:
@SimenB What you point out is correct, coverage collection behaves differently when using `-o`, but I think this is a good starting point to implement a feature that was honestly missing.

There are some candidates to these "default coverage filter" we're looking for.`Runtime.shouldInstrument` is a good one, however I've tried it and it didn't work. Anyways it is used because the coverage for each file is initialized using `generateEmptyCoverage` which correctly implements `Runtime.shouldInstrument`.

I think the main issue right now is that the way to filter the coverage for only changed files is to set `collectCoverageFrom` in the `globalConfig`, and alters some aspects of the coverage collection behavior as if it was the user who is forcing the collection of coverage from these files.

`jest-cli/runJest.js`
```javascript

let collectCoverageFrom = [];

...

if (matches.collectCoverageFrom) {
        collectCoverageFrom = ...
}

...

if (collectCoverageFrom.length) {
    const newConfig: GlobalConfig = {...globalConfig, collectCoverageFrom};
    globalConfig = Object.freeze(newConfig);
}
```

Maybe using a different option to filter coverage for only changed files would fix the problem and differentiate between user and internal filters.

MarcoScabbiolo commented:
> 

Indeed:

`jest-cli/reporters/coverage_reporter#_addUntestedFiles`

```javascript
if (
  globalConfig.collectCoverageFrom &&
  globalConfig.collectCoverageFrom.length
) {
  context.hasteFS
    .matchFilesWithGlob(globalConfig.collectCoverageFrom, config.rootDir)
    .forEach(filePath =>
      files.push({
        config,
        path: filePath,
      }),
    );
}
```

stipsan commented:
I'm sorry that I've not checked up on this until now, given that it's 20 days since I said I would look into it… But I'm really happy to see a bug report turned into a first time PR! 🎉 

>Maybe using a different option to filter coverage for only changed files would fix the problem and differentiate between user and internal filters.

I ran into this as well, it was one of the most tricky parts of implementing this to begin with 🤔

MarcoScabbiolo commented:
I finally got it, most of the previous code was deleted as it was unnecessary, I added `changedFiles` to `globalConfig` and then use it to exclude files from coverage, much clean and simpler.

I've deleted the `runJestWithCoverage` unit tests file as it tested only the previous behavior, using `collectCoverageFrom` for this same purpose. The e2e tests should cover these changes, but it would be nice to add new unit tests if somebody feels like it.

SimenB commented on packages/jest-runtime/src/shouldInstrument.js:
```diff
@@ -88,5 +88,12 @@ export default function shouldInstrument(
 
+  if (
+    options.changedFiles &&
+    !options.changedFiles.has(path.resolve(filename))
```
why we you call `path.resolve`?

MarcoScabbiolo replied:
Removed

SimenB commented on TestUtils.js:
```diff
@@ -13,6 +13,7 @@ import type {GlobalConfig, ProjectConfig} from 'types/Config';
 
 const DEFAULT_GLOBAL_CONFIG: GlobalConfig = {
   bail: 0,
+  changedFiles: null,
```
I agree it's practical in order to propagate, but this should not be part of the config (global or project). The value has to be manually passed through

MarcoScabbiolo replied:
Deleted

SimenB commented on packages/jest-cli/src/runJest.js:
```diff
@@ -147,57 +147,32 @@ export default (async function runJest({
+      ...globalConfig,
+      changedFiles:
+        changedFiles &&
+        new Set(Array.from(changedFiles).map(p => path.resolve(p))),
```
why the `resolve` call?

MarcoScabbiolo replied:
Removed

SimenB commented:
>since that was being done in the previous implementation I figured it would be ok

That was done mostly since `collectCoverageFrom` is a config option already. It was a hacky solution when added and we should have caught it then. So this is a really good cleanup 🙂 

>I was using path.resolve to make sure filenames are normalized to absolute ones.

That seems fine, but they should resolve from something, e.g. `rootDir`. We don't want to use `cwd`, which is what `path.resolve` falls back to. If we don't need the `resolve` call though, that'd be the best

MarcoScabbiolo commented:
I've refactored the `-o`/`--coverage` implementation based on feedback by @SimenB 
Thank you for your assistance and for maintaining this awesome project.

SimenB commented on packages/jest-cli/src/TestScheduler.js:
```diff
@@ -173,6 +174,7 @@ export default class TestScheduler {
         // $FlowFixMe
         testRunners[config.runner] = new (require(config.runner): TestRunner)(
           this._globalConfig,
+          this._options.changedFiles,
```
can you put this in an object? `type RunData = { changedFiles: ?Set<Path> }` or something like that

SimenB replied:
maybe put `changedFiles` into `TestSchedulerContext` instead, and pass that whole thing?

SimenB commented on packages/jest-cli/src/TestScheduler.js:
```diff
@@ -262,7 +264,9 @@ export default class TestScheduler {
     if (!isDefault && collectCoverage) {
-      this.addReporter(new CoverageReporter(this._globalConfig));
+      this.addReporter(
+        new CoverageReporter(this._globalConfig, this._options.changedFiles),
```
Pass in an object

SimenB commented on packages/jest-cli/src/TestScheduler.js:
```diff
@@ -288,7 +292,9 @@ export default class TestScheduler {
     if (collectCoverage) {
-      this.addReporter(new CoverageReporter(this._globalConfig));
+      this.addReporter(
+        new CoverageReporter(this._globalConfig, this._options.changedFiles),
```
Pass in an object

jeysal commented on packages/jest-cli/src/SearchSource.js:
```diff
@@ -158,29 +158,48 @@ export default class SearchSource {
+        testModulesMap.reduce((acc, testModule) => {
+          if (testModule.dependencies) {
+            testModule.dependencies
+              .filter(p => allPathsAbsolute.some(ap => ap === p))
```
`allPathsAbsolute.includes(p)`

jeysal commented on packages/jest-cli/src/SearchSource.js:
```diff
@@ -158,29 +158,48 @@ export default class SearchSource {
+          if (testModule.dependencies) {
+            testModule.dependencies
+              .filter(p => allPathsAbsolute.some(ap => ap === p))
+              .forEach(p => acc.add(p));
```
Maybe it's just me, but I find mutating the accumulated value in `reduce` super weird. Can we maybe move all the processing out of `return` and just `forEach` over `testModulesMap`?

MarcoScabbiolo replied:
Mutating the accumulated is a common use case of `reduce`, do you know of any drawbacks to it other than "finding it weird"?

I can create a variable `collectCoverageFrom` and then assign it to the result, but it would be an unnecessary variable to create as the values can be converted in a single expression.

jeysal replied:
Is it? I've never seen it used that way.
The concept comes from functional programming, the normal way to use it is combining the `acc` and `next` into a new value and returning that.
I also think in general it wouldn't hurt to break up that enormous `return` statement?

jeysal commented on packages/jest-cli/src/runJest.js:
```diff
@@ -147,57 +147,21 @@ export default (async function runJest({
         context,
         outputStream,
-        changedFilesPromise,
+        changedFilesPromise && (await changedFilesPromise),
```
@SimenB think this will affect startup perf?

SimenB commented:
Will this fix #7153?

MarcoScabbiolo commented:
Yes @SimenB, that PR aims to tackle the same underlying issue

SimenB commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@
 - `[jest-config]` Make sure `normalize` can consume `Defaults` without warnings ([#7742](https://github.com/facebook/jest/pull/7742))
 - `[jest-config]` Allow `moduleFileExtensions` without 'js' for custom runners ([#7751](https://github.com/facebook/jest/pull/7751))
 - `[jest-cli]` Load transformers before installing require hooks ([#7752](https://github.com/facebook/jest/pull/7752)
+- `[jest-cli]` Refactor `-o` and `--coverage` combined ([#7611](https://github.com/facebook/jest/pull/7611))
```
can you merge in/rebase on `master`, then move this entry up under `master`?

SimenB commented:
Thank you so much!
